File: OptimoveSDK/optimove-sdk/src/main/java/com/optimove/android/gamifywidgetsdk/AndroidBridge.java
package com.optimove.android.gamifywidgetsdk;

import android.util.Log;
import android.webkit.JavascriptInterface;

import androidx.annotation.Nullable;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * Native bridge exposed to the widget's JavaScript as {@code window.AndroidBridge}.
 *
 * <p>The widget calls:
 * <ul>
 *   <li>{@code window.AndroidBridge.closeWidget()} — dismiss the dialog</li>
 *   <li>{@code window.AndroidBridge.receiveMessage(json)} — widget → SDK messages,
 *       including READY (loyalty handshake) and CLOSE</li>
 * </ul>
 */
class AndroidBridge {

    private final Runnable onClose;
    @Nullable private final Runnable onReady;

    AndroidBridge(Runnable onClose, @Nullable Runnable onReady) {
        this.onClose = onClose;
        this.onReady = onReady;
    }

    @JavascriptInterface
    public void closeWidget() {
        onClose.run();
    }

    @JavascriptInterface
    public void receiveMessage(String json) {
        try {
            String type = new JSONObject(json).getString("type");
            if ("READY".equals(type)) {
                if (onReady != null) {
                    onReady.run();
                }
            } else if ("CLOSE".equals(type)) {
                onClose.run();
            }
        } catch (JSONException e) {
            Log.d(GamifyWidgetSDK.TAG, "Incorrect message format: " + json);
        }
    }
}

File: OptimoveSDK/gamify-widget-sdk/src/main/java/com/optimove/android/gamifywidgetsdk/AndroidBridge.kt
package com.optimove.android.gamifywidgetsdk

import android.util.Log
import android.webkit.JavascriptInterface
import org.json.JSONException
import org.json.JSONObject

/**
 * Native bridge exposed to the widget's JavaScript as `window.AndroidBridge`.
 *
 * The widget calls:
 *   window.AndroidBridge.closeWidget()        — to dismiss the bottom sheet
 *   window.AndroidBridge.receiveMessage(json) — generic widget → SDK messages,
 *       including the READY handshake (type = "READY")
 */
internal class AndroidBridge(
    private val onClose: () -> Unit,
    private val onReady: () -> Unit
) {

    @JavascriptInterface
    fun closeWidget() {
        onClose()
    }

    @JavascriptInterface
    fun receiveMessage(json: String) {
        try {
            if (JSONObject(json).getString("type") == "READY") {
                onReady()
            }
        } catch (e: JSONException) {
            Log.d(TAG, "Incorrect message format: $json")
        }
    }

    companion object {
        private const val TAG = "GamifyWidget"
    }
}
